server/src/main/java/org/elasticsearch/common/collect/Iterators.java

```diff
@@ -43,44 +47,45 @@ public static <T> Iterator<T> concat(Iterator<? extends T>... iterators) {
             throw new NullPointerException("iterators");
         }
 
-        // explicit generic type argument needed for type inference
-        return new ConcatenatedIterator<T>(iterators);
+        for (int i = 0; i < iterators.length; i++) {
+            if (iterators[i].hasNext()) {
+                // explicit generic type argument needed for type inference
+                return new ConcatenatedIterator<T>(iterators, i);
+            }
+        }
+
+        return Collections.emptyIterator();
     }
 
-    static class ConcatenatedIterator<T> implements Iterator<T> {
+    private static class ConcatenatedIterator<T> implements Iterator<T> {
         private final Iterator<? extends T>[] iterators;
-        private int index = 0;
+        private int index;
 
-        @SafeVarargs
-        @SuppressWarnings("varargs")
-        ConcatenatedIterator(Iterator<? extends T>... iterators) {
-            if (iterators == null) {
-                throw new NullPointerException("iterators");
-            }
-            for (int i = 0; i < iterators.length; i++) {
+        ConcatenatedIterator(Iterator<? extends T>[] iterators, int startIndex) {
+            for (int i = startIndex; i < iterators.length; i++) {
                 if (iterators[i] == null) {
                     throw new NullPointerException("iterators[" + i + "]");
                 }
             }
             this.iterators = iterators;
+            this.index = startIndex;
         }
 
         @Override
         public boolean hasNext() {
-            boolean hasNext = false;
-            while (index < iterators.length && (hasNext = iterators[index].hasNext()) == false) {
-                index++;
```

Comment:
I find it weird to be changing state in `hasNext()` like this, I prefer to do all the moving in `next()` so that we're always pointing at a nonempty iterator until done.

Reply:
++